resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,265 @@
+     - each `B0i` and `B1i` are equal types (syntactically)
+     - `R0` is **DOWN**(`T0`, `T1`)
+     - `B2i` is `B0i`
+     - `P2i` is **UP**(`P0i`, `P1i`)
+     - `Named2` contains `R2i xi` for each `xi` in both `Named0` and `Named1`
+        - where `R0i xi` is in `Named0`
+        - where `R1i xi` is in `Named1`
+        - and `R2i` is **UP**(`R0i`, `R1i`)
+     - `Named2` contains `R0i xi` for each `xi` in  `Named0` and not `Named1`
+     - `Named2` contains `R1i xi` for each `xi` in  `Named1` and not `Named0`
+
+- **DOWN**(`T Function<...>(...)`, `S Function<...>(...)`) = `bottom` otherwise
+
+
+- **DOWN**(`T1`, `T2`) = `T1` if `T1` <: `T2`
+- **DOWN**(`T1`, `T2`) = `T2` if `T2` <: `T1`
+- **DOWN**(`T1`, `T2`) = `bottom` otherwise
+
+
+## Issues and Interesting examples
+
+### Type variable bounds
+
+The definition of upper bound for type variables does not guarantee termination.
+Counterexample:
+
+```dart
+void foo<T extends List<S>, S extends List<T>>() {
+  T x;
+  S y;
+  var a = (x == y) ? x : y;
+}
+```
+
+It should be changed to close the bound with respect to all of the type
+variables declared in the same scope, using the greatest closure definition.
+
+### Generic functions
+
+The CFE currently implements upper bounds for generic functions incorrectly. Example:
+
+```dart
+typedef G0 = T Function<T>(T x);
+typedef G1 = T Function<T>(T x);
+void main() {
+   {
+    G0 x;
+    G1 y;
+    // Analyzer: T Function<T>(T)
+    // CFE: bottom -> Object
+    var a = (x == y) ? x : y;
+
+    }
+}
+```
+
+Both the CFE and the analyzer currently implement lower bounds for generic
+functions incorrectly.  Example:
+
+```dart
+typedef G0 = T Function<T>(T x);
+typedef G1 = T Function<T>(T x);
+void main() {
+  {
+    void Function(G0) x;
+    void Function(G1) y;
+    int z;
+    // Analyzer: void Function(Never Function(Object))
+    // CFE:      void Function(bottom-type Function(Object))
+    var a = (x == y) ? x : y;
+  }
+}
+```
+
+## Asymmetry
+
+The current algorithm is asymmetric.  There is an equivalence class of top
+types, and we correctly choose a canonical representative for bare top types
+using the **MORETOP** predicate.  However, when two different top types are
+embedded in two mutual subtypes, we don't correctly choose a canonical
+representative.
+
+```dart
+import 'dart:async';
+
+void main () {
+  List<FutureOr<Object>> x;
+  List<dynamic> y;
+  String s;
+  // List<dynamic>
+  var a = (x == y) ? x : y;
+  // List<FutureOr<Object>>
+  var b = (x == y) ? y : x;
+```
+
+The best solution for this is probably to normalize the types.  This is fairly
+straightforward: we just normalize `FutureOr<T>` to the normal form of `T` when
+`T` is a top type.  We can then inductively apply this across the rest of the
+types.  Then, whenever we have mutual subtypes, we just return the normal form.
+This would be breaking, albeit hopefully only in a minor way.
```

Comment:
It's breaking in the sense that it changes the MORETOP order (using the abbreviation `F<..>` for `FutureOr<..>`, `F2<..>` for `FutureOr<FutureOr<..>>` etc):

`void > dynamic > Object > F<void> > F<dynamic> > F<Object> > F2<void> > ..`

becomes `void > dynamic > Object`, folding `Fk<T>` into `T` for all k, for `T` in {`void`, `dynamic`, `Object`}. For example, a raw `FutureOr` could introduce `dynamic` into a signature (say, as the return type), where we would previously have `Object`.